app/browser/api/ledger.js

```diff
@@ -2627,6 +2725,46 @@ const onPublisherTimestamp = (state, oldTimestamp, newTimestamp) => {
   })
 }
 
+const checkReferralActivity = (state) => {
+  const downloadId = updateState.getUpdateProp(state, 'downloadId')
+
+  if (!downloadId) {
+    updater.checkForUpdate(false, true)
+    return
+  }
+
+  module.exports.roundtrip({
+    server: referralServer,
+    method: 'PUT',
+    path: '/promo/activity',
+    payload: {
+      download_id: downloadId,
+      api_key: referralAPI
+    }
+  }, {}, activityRoundTrip)
+}
+
+const activityRoundTrip = (err, response, body) => {
+  if (err) {
+    if (clientOptions.verboseP) {
+      console.error(err)
+    }
+    updater.checkForUpdate(false, true)
+    return
+  }
+
+  if (body && body.finalized) {
+    appActions.onReferralActivity()
+    return
+  }
+
+  if (clientOptions.verboseP) {
+    console.log('Referral is still not ready, please wait at least 30days')
+  }
+
+  updater.checkForUpdate(false, true)
```

Comment:
minor: `activityRoundTrip` and `checkReferralActivity` would be clearer if they only handled things related to referral activity instead of also checking for browser updates. instead of returning after `appActions.checkReferralActivity()` in `updater.checkForUpdate`, you could always check for the browser update so that these functions here don't need to check for browser updates.

Reply:
we want to reduce amount of calls, so we only want to call update server when referral fails.